Manage grafana-agent extra config and secret

marieroque opened the pull request:
towards https://github.com/giantswarm/giantswarm/issues/28486

To configure properly `grafana-agent` to be able to scrape Kubernetes Events, we need to add a `river` config to `grafana-agent`.
This configuration is set in an extra config => <cluster_id>-grafana-agent-config
This configuration needs secret values like credentials to push data on `loki`. 
To store those secret values, we create an extra secret => <cluster_id>-grafana-agent-secret
The `river` configuration use those values as environment variables pass through `envFrom` property.

QuentinBisson commented on pkg/resource/apps-toggle/observability_bundle_configmap.go:
```diff
@@ -19,16 +19,20 @@ type app struct {
+			promtailName: {
+				Enabled: true,
+			},
+			grafanaAgentName: {
```
I think you Can use the string without using a variable ;)

QuentinBisson commented on pkg/resource/apps-toggle/reconciler.go:
```diff
@@ -40,31 +40,31 @@ func (r *Reconciler) GetObservabilityBundleVersion(ctx context.Context, lc logge
 
-	// Get desired configmap to enable promtail.
+	// Get desired configmap to enable Promtail/GrafanaAgent.
 	desiredConfigMap, err := GenerateObservabilityBundleConfigMap(lc, observabilityBundleVersion)
```
I would replace all Promtail/GrafanaAgent occurrences by logging agents tbh because if we add or remove an app it will bé a pain

QuentinBisson commented on pkg/resource/apps-toggle/reconciler.go:
```diff
@@ -21,7 +21,7 @@ import (
 
 // Reconciler implements a reconciler.Interface to handle
-// Promtail and GrafanaAgent toggle: enable or disable Promtail/GrafanaAgent in a given Cluster.
+// Logging agents toggle: enable or disable logging agents in a given Cluster.
```
Maybe we should rename the resource to logging-agent-toggle or agent-toggle?

QuentinBisson commented on pkg/resource/promtail-client/promtail-client.go:
```diff
@@ -154,8 +154,8 @@ func GetPromtailCredentials(lc loggedcluster.Interface, credentialsSecret *v1.Se
 	var credentials = make(map[string]string)
-	println(promtailYaml.Config.Clients)
-	println(promtailYaml.Config.Clients[0].URL)
+	println(promtailYaml)
```
Why do we even need a println?

marieroque replied:
It's a draft... not ready to review!
You're right, I'll do my debug test locally and not push them from now on :-)

QuentinBisson replied:
Ah my bad :) but we had println before 😂

QuentinBisson commented on CHANGELOG.md:
```diff
@@ -7,6 +7,10 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
+### Added
+
+- Configure grafana-agent config.
```
I would add a bit more to this to explain why we even add grafana-agent config, WDYT?

QuentinBisson commented on README.md:
```diff
@@ -28,12 +31,14 @@ The operator is built around a central reconciler, that calls multiple sub-recon
 * Promtail config setups some Promtail settings (like which logs to collect)
-* Promtail toggle enables/disables promtail deployment on WCs
 * Promtail wiring ensures promtail-app reads configs from previous steps
+* Grafana-Agent config setups some Grafana-Agent settings (like credentials to access loki write, configuration to scrape Kubernetes Events)
```
```suggestion
* grafana-agent config setups some grafana-agent settings (like credentials to access loki write, configuration to scrape Kubernetes Events)
```

QuentinBisson commented on pkg/resource/grafana-agent-config/reconciler.go:
```diff
@@ -0,0 +1,95 @@
+		return ctrl.Result{}, errors.WithStack(err)
+	}
+
+	promtailCredentials, err := promtailclient.GetPromtailCredentials(lc, &promtailSecret)
```
Maybe we should rename promtailCredentials to loki-client-credentials or logging-agent-credentials?

QuentinBisson replied:
I think we should check all things that are named promtail even :)

QuentinBisson commented on pkg/resource/grafana-agent-config/grafana-agent-config.go:
```diff
@@ -0,0 +1,122 @@
+
+// GenerateGrafanaAgentConfig returns a configmap for
+// the grafana-agent extra-config
+func GenerateGrafanaAgentConfig(lc loggedcluster.Interface, promtailCredentials map[string]string) (v1.ConfigMap, error) {
```
```suggestion
func GenerateGrafanaAgentConfig(lc loggedcluster.Interface, credentials map[string]string) (v1.ConfigMap, error) {
```

marieroque commented:
Tested on `gaia` MC ✅ 
Tested on `gaia` WC 🔴 => errors in `grafana-agent`
```
ts=2023-10-26T11:11:00.241793779Z level=error msg="event watcher exited with error" component=loki.source.kubernetes_events.local err="creating informers cache: could not create RESTMapper from config: Get \"https://172.31.0.1:443/api\": dial tcp 172.31.0.1:443: i/o timeout"
```

QuentinBisson commented on pkg/resource/grafana-agent-config/reconciler.go:
```diff
@@ -27,8 +29,18 @@ func (r *Reconciler) ReconcileCreate(ctx context.Context, lc loggedcluster.Inter
+
 	// Get desired config
-	desiredGrafanaAgentConfig, err := GenerateGrafanaAgentConfig(lc)
+	desiredGrafanaAgentConfig, err := GenerateGrafanaAgentConfig(lc, currentApp.Spec.Namespace)
```
How is that fixin it ?

marieroque replied:
It retrieves the `grafana-agent` App namespace in the App spec (here "kube-system")

QuentinBisson replied:
Ah ok :)

QuentinBisson commented on pkg/resource/grafana-agent-config/grafana-agent-config.go:
```diff
@@ -0,0 +1,110 @@
+// the grafana-agent extra-config
+func GenerateGrafanaAgentConfig(lc loggedcluster.Interface, appNamespace string) (v1.ConfigMap, error) {
+
+	namespacesScraped := "[]"
```
```suggestion
	scrapedNamespaces := "[]"
```

marieroque commented:
After adding netpol, issue on WC is resolved.

Tested on `gaia` WC ✅

QuentinBisson commented on pkg/resource/logging-agents-toggle/observability_bundle_configmap.go:
```diff
@@ -19,16 +19,19 @@ type app struct {
+// the observabilitybundle application to enable logging agents.
 func GenerateObservabilityBundleConfigMap(lc loggedcluster.Interface, observabilityBundleVersion semver.Version) (v1.ConfigMap, error) {
-	appName := "promtail"
+	promtailName := "promtail"
```
```suggestion
	promtailAppName := "promtail"
```

QuentinBisson commented:
It looks alright at a first glance but could you split this into smaller PRs so it's easier to review for my tired eyes ?

Maybe start by moving ReadLokiIngress into common and rename to logging agents  and then the rest ?

marieroque commented:
> It looks alright at a first glance but could you split this into smaller PRs so it's easier to review for my tired eyes ?
> 
> Maybe start by moving ReadLokiIngress into common and rename to logging agents and then the rest ?

Well, I moved ReadLokiIngress into common when I added code for grafana-agent to avoid duplicate code.
So all is linked...

QuentinBisson commented:
I know it's linked :)

marieroque commented:
You're nasty if you want me to separate in PR anyway...

QuentinBisson commented:
Nasty ? You guys ask me that all the time xD But sure i'll try to review that

QuentinBisson commented on pkg/common/common.go:
```diff
@@ -1,11 +1,29 @@
+	grafanaAgentResourceName = "grafana-agent-secret"
+)
+
+func GetGrafanaAgentResourceName() string {
```
```suggestion
func GrafanaAgentSecretName() string {
```

And why does it need to be in common when it could be here instead https://github.com/giantswarm/logging-operator/blob/508e156efd552cf2004e1cddd12d0c9bb65223e0/pkg/resource/grafana-agent-secret/grafana-agent-secret.go ?

marieroque replied:
because I use it here also
https://github.com/giantswarm/logging-operator/blob/508e156efd552cf2004e1cddd12d0c9bb65223e0/pkg/resource/grafana-agent-config/grafana-agent-config.go#L74

QuentinBisson replied:
Oh I missed that then a renaming is enough :)

QuentinBisson commented on pkg/common/common.go:
```diff
@@ -22,3 +40,19 @@ func AddCommonLabels(labels map[string]string) {
+
+// Read Loki URL from ingress
+func ReadLokiIngressURL(ctx context.Context, lc loggedcluster.Interface, client client.Client) (string, error) {
+
```
```suggestion
```

QuentinBisson commented on pkg/resource/grafana-agent-config/grafana-agent-config.go:
```diff
@@ -0,0 +1,110 @@
+// GenerateGrafanaAgentConfig returns a configmap for
+// the grafana-agent extra-config
+func GenerateGrafanaAgentConfig(lc loggedcluster.Interface, appNamespace string) (v1.ConfigMap, error) {
+
```
```suggestion
```

QuentinBisson commented on pkg/resource/grafana-agent-config/grafana-agent-config.go:
```diff
@@ -0,0 +1,110 @@
+	name = "` + fmt.Sprintf("%s-%s", lc.GetClusterName(), common.GetGrafanaAgentResourceName()) + `"
+}
+
+loki.write "default" {
```
Damn I hate multiline strings

QuentinBisson commented on pkg/resource/grafana-agent-secret/reconciler.go:
```diff
@@ -0,0 +1,95 @@
+
+	// Retrieve secret containing credentials
+	var loggingCredentialsSecret v1.Secret
+	err := r.Client.Get(ctx, types.NamespacedName{Name: loggingcredentials.LoggingCredentialsSecretMeta(lc).Name, Namespace: loggingcredentials.LoggingCredentialsSecretMeta(lc).Namespace},
```
I think you can use client.ObjectKeyFromObject{loggingcredentials.LoggingCredentialsSecretMeta(lc)} instead of the whole types.NamespacedName{Name: loggingcredentials.LoggingCredentialsSecretMeta(lc).Name, Namespace: loggingcredentials.LoggingCredentialsSecretMeta(lc).Namespace}

marieroque replied:
I got error
```
cannot use loggingcredentials.LoggingCredentialsSecretMeta(lc) (value of type "k8s.io/apimachinery/pkg/apis/meta/v1".ObjectMeta) as client.Object value in argument to client.ObjectKeyFromObject: "k8s.io/apimachinery/pkg/apis/meta/v1".ObjectMeta does not implement client.Object (missing method DeepCopyObject)compiler[InvalidIfaceAssign](https://pkg.go.dev/golang.org/x/tools/internal/typesinternal#InvalidIfaceAssign)
```

QuentinBisson replied:
Alright

marieroque commented:
> Nasty ? You guys ask me that all the time xD But sure i'll try to review that

Only 12 files changed... little PR actually ;-)

QuentinBisson commented on pkg/resource/logging-agents-toggle/observability_bundle_configmap.go:
```diff
@@ -19,16 +19,19 @@ type app struct {
+			promtailAppName: {
+				Enabled: true,
+			},
+			"grafanaAgent": {
```
We need to use an observability-bundle version toggle here

QuentinBisson replied:
https://gigantic.slack.com/archives/C01176DKNP4/p1699452290447569

marieroque replied:
I did it, it is coming...

QuentinBisson commented on pkg/common/common.go:
```diff
@@ -21,7 +21,7 @@ const (
 )
 
-func GetGrafanaAgentResourceName() string {
+func GrafanaAgentResourceName() string {
```
```suggestion
func GrafanaAgentResourceName() string {
```

That's still a secret, not a resource

marieroque replied:
IIRC, when I set GrafanaAgentSecretName, I had validation issues because of `secret` in the name

QuentinBisson replied:
Sure but we can ignore that using something like https://github.com/giantswarm/logging-operator/blob/7715f96f71bfed748b5ff465e59495a6acf36c6c/pkg/resource/loki-auth/loki-auth.go#L22
